File: aoi/dnd/grader.cpp
#include "dnd.h"
#include <cstdio>
#include <cstdlib>
#include <vector>
#include <utility>
#include <csetjmp>

using namespace std;

static const int MAX_N = 1005;
static const int MAX_NOTE = 42;
static const int MAX_STEP = 4000000;

#ifdef LOG
	#define log(...) fprintf(stderr,__VA_ARGS__)
#else
	#define log(...)
#endif

static int n;
static vector<pair<int,int>> adjl[MAX_N];
static bool d[MAX_N];
static jmp_buf restart;
static int p,steps;
static string note[MAX_N];

static void raise(string e)
{
	fprintf(stderr,"%s\n",e.c_str());
	exit(1);
}

int numOfDoors() { return adjl[p].size(); }

int takeDoor(int k)
{
	if (k<=0||k>adjl[p].size())
		raise("Diese Tuer gibt es nicht");
	steps++;
	if (steps>MAX_STEP)
		raise("Zu viele Schritte");
	auto e=adjl[p][k-1];
	log("Von %d zu %d\n",p,e.first);
	p=e.first;
	if (p==n)
	{
		printf("Du bist entkommen, %d Schritte\n",steps);
		exit(0);
	}
	if (d[p])
	{
		log("Tod\n");
		longjmp(restart,1);
	}
	log("Bei %d\n",p);
	return e.second;
}

void takeNote(string s)
{
	if (s.size()>MAX_NOTE)
		raise("Notiz zu lang");
	note[p]=s;
}

string readNote() { return note[p]; }

int main()
{
	int m,a,b;
	scanf("%d %d",&n,&m);
	while (m--)
	{
		scanf("%d %d",&a,&b);
		adjl[a].push_back(make_pair(b,adjl[b].size()+1));
		adjl[b].push_back(make_pair(a,adjl[a].size()));
	}
	scanf("%d",&m);
	while (m--)
	{
		scanf("%d",&a);
		d[a]=true;
	}
	setjmp(restart);
	p=1;
	log("Bei 1\n");
	findExit();
	raise("Ausgang nicht gefunden");
	return 0;
}

File: aoi/dnd/grader.java
import java.util.Scanner;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;

class grader
{
    static final int MAX_N = 1005;
    static final int MAX_NOTE = 42;
    static final int MAX_STEP = 4000000;

    static final int AC = 0;
    static final int INVALID_DOOR = 1;
    static final int TOO_MANY_STEPS = 2;
    static final int NOTE_TOO_LONG = 3;
    static final int NOT_FOUND = 4;
    static final int EXCEPTION_CAUGHT = 5;

    static class Edge
    {
        public int to;
        public int otherDoor;
        Edge(int t, int d) { to = t; otherDoor = d; }
    }

    static int n;
    static ArrayList<ArrayList<Edge>> adjl;
    static boolean[] d;
    static boolean currentlyDead = false;

    static class State
    {
        public int position = 1;
        public int steps = 0;
        public String[] notes = new String[n];
    }
    static State st;

    static void judge(int status)
    {
        switch(status)
        {
            case AC:
                System.out.println("Du bist entkommen, " + st.steps + " Schritte");
                break;
            case INVALID_DOOR:
                System.out.println("Diese Tuer gibt es nicht");
                break;
            case TOO_MANY_STEPS:
                System.out.println("Zu viele Schritte");
                break;
            case NOTE_TOO_LONG:
                System.out.println("Notiz zu lang");
                break;
            case NOT_FOUND:
                System.out.println("Ausgang nicht gefunden");
                break;
            case EXCEPTION_CAUGHT:
                System.out.println("DeathException darf nicht gefangen werden");
                break;
        }
        System.exit(0);
    }

    static public int numOfDoors()
    {
        return adjl.get(st.position).size();
    }

    static public int takeDoor(int k)
    {
        if(currentlyDead)
            judge(EXCEPTION_CAUGHT); // someone caught the exception?

        if (k <= 0 || k > adjl.get(st.position).size())
            judge(INVALID_DOOR);
        st.steps++;
        if (st.steps > MAX_STEP)
            judge(TOO_MANY_STEPS);
        Edge e = adjl.get(st.position).get(k-1);
        st.position = e.to;
        if (st.position == n)
        {
            judge(AC);
        }
        if (d[st.position])
        {
            currentlyDead = true;
            throw new DeathException();
        }
        return e.otherDoor;
    }
    
    static public void takeNote(String s)
    {
        if (s.length() > MAX_NOTE)
		    judge(NOTE_TOO_LONG);
	    st.notes[st.position] = s;
    }

    
    static public String readNote()
    {
        String s = st.notes[st.position];
        return s != null ? s : "";
    }

    static class DeathException extends RuntimeException {}

    public static void main(String[] args)
    {
        Scanner in = new Scanner(new BufferedReader(new InputStreamReader(System.in)));
        n = in.nextInt();
        adjl = new ArrayList<ArrayList<Edge>>();
        adjl.add(null);
        for (int i = 0; i < n; i++)
            adjl.add(new ArrayList<Edge>());
        d = new boolean[n+1];
        for (int m = in.nextInt(); m > 0; m--)
        {
            int a = in.nextInt();
            int b = in.nextInt();
            adjl.get(a).add(new Edge(b, adjl.get(b).size() + 1));
            adjl.get(b).add(new Edge(a, adjl.get(a).size()));
        }
        for (int m = in.nextInt(); m > 0; m--)
        {
            d[in.nextInt()] = true;
        }

        st = new State();
        for(;;)
        {
            try
            {
                st.position = 1;
                new DND().findExit();
                judge(NOT_FOUND);
            }
            catch (DeathException e)
            {
                currentlyDead = false;
            }
        }
    }
}
